engine/launcher/src/main/scala/org/enso/launcher/installation/DistributionManager.scala

```diff
@@ -1,7 +1,267 @@
+
+  /**
+    * Specifies whether the launcher has been run as a portable distribution or
+    * it is a locally installed distribution.
+    */
+  lazy val isRunningPortable: Boolean = {
+    val portable = detectPortable()
+    Logger.debug(s"Launcher portable mode = $portable")
+    if (portable && LocallyInstalledDirectories.installedDistributionExists) {
+      val installedRoot   = LocallyInstalledDirectories.dataDirectory
+      val installedBinary = LocallyInstalledDirectories.binaryExecutable
+
+      Logger.debug(
+        s"The launcher is run in portable mode, but an installed distribution" +
+        s" is available at $installedRoot."
+      )
+
+      if (Files.exists(installedBinary)) {
+        if (installedBinary == env.getPathToRunningExecutable) {
+          Logger.debug(
+            "That distribution seems to be corresponding to this launcher " +
+            "executable, that is running in portable mode."
+          )
+        } else {
+          Logger.debug(
+            s"However, that installed distribution most likely uses another " +
+            s"launcher executable, located at $installedBinary."
+          )
+        }
+      }
+    }
+    portable
+  }
+
+  /**
+    * Determines paths that should be used by the launcher.
+    */
+  lazy val paths: DistributionPaths = {
+    val paths = detectPaths()
+    Logger.debug(s"Detected paths are: $paths")
+
+    paths
+  }
+
+  private val PORTABLE_MARK_FILENAME = ".enso.portable"
+  val ENGINES_DIRECTORY              = "dist"
+  val RUNTIMES_DIRECTORY             = "runtime"
+  val CONFIG_DIRECTORY               = "config"
+  val BIN_DIRECTORY                  = "bin"
+  private val TMP_DIRECTORY          = "tmp"
+
+  private def detectPortable(): Boolean = Files.exists(portableMarkFilePath)
+  private def possiblePortableRoot: Path =
+    env.getPathToRunningExecutable.getParent.getParent
+
+  private def portableMarkFilePath: Path =
+    possiblePortableRoot / PORTABLE_MARK_FILENAME
+
+  private def detectPaths(): DistributionPaths =
+    if (isRunningPortable) {
+      val root = env.getPathToRunningExecutable.getParent.getParent
+      DistributionPaths(
+        dataRoot = root,
+        runtimes = root / RUNTIMES_DIRECTORY,
+        engines  = root / ENGINES_DIRECTORY,
+        config   = root / CONFIG_DIRECTORY,
+        tmp      = root / TMP_DIRECTORY
+      )
+    } else {
+      val dataRoot   = LocallyInstalledDirectories.dataDirectory
+      val configRoot = LocallyInstalledDirectories.configDirectory
+      DistributionPaths(
+        dataRoot = dataRoot,
+        runtimes = dataRoot / RUNTIMES_DIRECTORY,
+        engines  = dataRoot / ENGINES_DIRECTORY,
+        config   = configRoot,
+        tmp      = dataRoot / TMP_DIRECTORY
+      )
+    }
+
+  /**
+    * A helper for managing directories of the local installation.
+    *
+    * It returns paths of the local installation even if the launcher is running
```

Comment:
"Local installation" is ambiguous. I assume that you mean just "the non-portable installation".

Reply:
I couldn't come up with a good name that is well descriptive for this concept. I guess "non-portable installation" is more understandable.